Adopt C++ Concepts in wtf/HashCountedSet

iidmsa opened the pull request:
#### 01eb3f483634081416ddc701fdf9ed0c6ab76c7e
<pre>
Adopt C++ Concepts in wtf/HashCountedSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=295789">https://bugs.webkit.org/show_bug.cgi?id=295789</a>
<a href="https://rdar.apple.com/problem/155626637">rdar://problem/155626637</a>

Reviewed by Darin Adler.

Replace std::enable_if with C++20 requires clauses for template constraints.

* Source/WTF/wtf/GetPtr.h:
* Source/WTF/wtf/HashCountedSet.h:
(WTF::Traits&gt;::find):
(WTF::Traits&gt;::find const):
(WTF::Traits&gt;::contains const):
(WTF::Traits&gt;::count const):
(WTF::Traits&gt;::remove):

Canonical link: <a href="https://commits.webkit.org/297305@main">https://commits.webkit.org/297305@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/069cf92d96f8529ea760cf5022c35f0a94efc4a5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111169 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30835 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21350 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117200 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61437 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31516 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39417 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84511 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114116 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25177 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100144 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64957 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24523 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18284 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61020 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103722 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94558 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18354 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120211 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109722 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38218 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28401 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93446 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38594 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96421 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93271 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23783 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38357 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16162 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34203 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38107 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43584 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133998 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37772 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36152 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41105 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39474 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/GetPtr.h:
```diff
@@ -15,7 +15,6 @@
  *  along with this library; see the file COPYING.LIB.  If not, write to
  *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
  *  Boston, MA 02110-1301, USA.
- *
```
Seems OK, but why did you include this?

iidmsa replied:
Just to make it consistent with other files. I just noticed most of the time there is not a blank line at the end.

darinadler commented on Source/WTF/wtf/GetPtr.h:
```diff
@@ -34,6 +33,12 @@ template <typename T> struct IsSmartPtr {
 };
 
+template<typename T>
+inline constexpr bool IsSmartPtr_v = IsSmartPtr<T>::value;
```
Great idea, not sure the naming is ideal with the mix of `InterCaps` and `underscore_separated_words`.

iidmsa replied:
Hmm, can we do `is_smart_ptr_v`?

darinadler replied:
I think that’s a big change in naming strategy for the WebKit project. For the most part our own identifiers don’t follow the standard library style, otherwise the class would be named `is_smart_ptr` already.

iidmsa replied:
Gotcha. Will resort to `IsSmartPtrV`

darinadler commented on Source/WTF/wtf/GetPtr.h:
```diff
@@ -34,6 +33,12 @@ template <typename T> struct IsSmartPtr {
+inline constexpr bool IsSmartPtr_v = IsSmartPtr<T>::value;
+
+template<typename T>
+inline constexpr bool IsSmartPtr_isNullable_v = IsSmartPtr<T>::isNullable;
```
Great idea, not sure the naming is ideal with the mix of `InterCaps` and `underscore_separated_words`.

iidmsa replied:
Hmm, can we do `is_smart_ptr_nullable_v`?

darinadler replied:
I think that’s a big change in naming strategy for the WebKit project. For the most part our own identifiers don’t follow the standard library style, otherwise the class would be named `is_smart_ptr` already.

iidmsa replied:
Gotcha. Will resort to `IsSmartPtrNullableV`

iidmsa committed:
Adopt C++ Concepts in wtf/HashCountedSet
https://bugs.webkit.org/show_bug.cgi?id=295789
rdar://problem/155626637

Reviewed by Darin Adler.

Replace std::enable_if with C++20 requires clauses for template constraints.

* Source/WTF/wtf/GetPtr.h:
* Source/WTF/wtf/HashCountedSet.h:
(WTF::Traits>::find):
(WTF::Traits>::find const):
(WTF::Traits>::contains const):
(WTF::Traits>::count const):
(WTF::Traits>::remove):

Canonical link: https://commits.webkit.org/297305@main
